python/pyarrow/_dataset.pyx

```diff
@@ -43,6 +44,51 @@ def _forbid_instantiation(klass, subclasses_instead=True):
     raise TypeError(msg)
 
 
+cdef class FileSource:
+
+    cdef:
+        CFileSource wrapped
+
+    def __cinit__(self, file, FileSystem filesystem=None):
+        cdef:
+            shared_ptr[CFileSystem] c_filesystem
+            c_string c_path
+            shared_ptr[CRandomAccessFile] c_file
+            shared_ptr[CBuffer] c_buffer
+
+        if isinstance(file, FileSource):
```

Comment:
This is odd. Should you have `_ensure_file_source` like other methods?

Reply:
Since we've decided not to export `FileSource` I think we can probably delete this class in a follow up (in favor of `cdef CFileSource _make_file_source(...)` or so)